### ISSUE TITLE
Thread-unsafe methods in Logging

### ISSUE THREAD

**Issue by perlun** [TwP/logging#129]
Hi,

Found this by accident:

``` ruby
require 'logging'

logger = Logging.logger['foo']
(0..10).map {
  Thread.new { logger.level = :info }
}.each(&:join)
```

The above program will produce the following error:

```
NameError: Undefined method info? for class '#<Class:0x15f57db3>'
                       (eval) at /Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/logger.rb:106
                  module_eval at org/jruby/RubyModule.java:2366
                   _meta_eval at /Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/logger.rb:420
           define_log_methods at /Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/logger.rb:105
                         each at org/jruby/RubyHash.java:1342
           define_log_methods at /Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/logger.rb:85
           define_log_methods at /Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/logger.rb:402
                       level= at /Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/logger.rb:291
                       (root) at ./foo.rb:7
                         call at org/jruby/RubyProc.java:271
  create_with_logging_context at /Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/diagnostic_context.rb:448
```

Of course, our _real_ use case isn't identical to this, but it illustrates the problem.

My problem is I am constructing a logger for a particular class. For _that instance_ of the class, I want to have a custom log level. In other words `Logging.logger('Some::Class')` isn't going to cut it, since I will always get the _same_ logger instance (which will trigger this error).

What is the recommended approach to resolving this? If I instead of using the class name change to just do a `to_s` of the class instance (which will contain a unique identifier), I get another error:

```
#<RuntimeError: can't add a new key into hash during iteration>
org/jruby/RubyHash.java:992:in `[]='
/Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/repository.rb:68:in `[]='
/Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/logger.rb:56:in `[]'
org/jruby/ext/thread/Mutex.java:149:in `synchronize'
/Volumes/extra/git/ecraft/ecraft.uxfactory.server-2014.1/src/server/gems/jruby/1.9/gems/logging-2.0.0/lib/logging/logger.rb:51:in `[]'
/Volumes/extra/git/ecraft/ecraft-web_api/lib/ecraft/web_api/log.rb:12:in `create_logger'
/Volumes/extra/git/ecraft/ecraft-web_api/lib/ecraft/web_api/monitor.rb:23:in `initialize'
```

That one wasn't as obvious to reproduce, but it's likely that it has a similar root cause (multiple loggers in this case being _registered_ into the repository in a reentrant way).

Any suggestion on how to fix this?

**Comment by perlun** [comment]
(It may be worth mentioning that this is a large, existing code base that I am porting over from its previous logging library, log4r.)

**Comment by TwP** [comment]
> My problem is I am constructing a logger for a particular class. For *that instance* of the class, I want to have a custom log level.

Each `Logging::Logger` created by the application is stored in the [`Logging::Repository`](https://github.com/TwP/logging/blob/master/lib/logging/repository.rb). This is what enables the framework to return the same `Logging::Logger` instance for any given class name.

The intent of the Logging framework is to have one Logger per class (as you have already observed). Creating a Logger for each **instance** of a class raises the concern of memory leaks. When the instance goes out of scope and is garbage collected, the Logger specific to that instance is retained by the `Logging::Repository` singleton. If you have long lived instances, then this is not a risk. If your instances are short lived, then I would advise against creating a Logger per instance.

But, if you really really really have to have a Logger per instance then here is what I'd do.

* each instance needs a unique, reproducible name
* use Ruby namespacing conventions for these names

```ruby
class ECraft::Unique
  def logger
    name = "#{self.class.name}::#{self.object_id}"
    Logging.logger[name]
  end
end
```

If your instances have human-friendly names, use that instead of the `self.object_id`. It will make life much easier when you are assigning log levels.

Using the Ruby namespace separator `::` is important! What this allows you to do is set a default log level on the `ECraft::Unique` logger. This log level will be inherited by all the instances. Changing the log level on the class will propagate that log level to all the instances. But you can still change individual instances thereby overriding the default from the class level.

```ruby
logger = Logging.logger["ECraft::Unique"]
logger.level = :warn

instance_logger = Logging.logger["ECraft::Unique::0x12345"]
instance_logger.level  #=> :warn
instance_logger.level = :debug
```

**Comment by TwP** [comment]
Regarding your first question - setting logging levels in a bunch of threads. Changing the level of a logger is a heavyweight method call. It will rewrite all the log level methods for that particular logger, and it will propagate the level change to all the children in the logging namespace hierarchy. For each child another heavyweight method call is made to rewrite the log level methods.

If we have a namespace such as `A::B::C::D::E`, then changing the level of the `A` logger will propagate to all the modules in the `A` namespace. This results in changing the log level of all 5 of these loggers:

* `A`
* `A::B`
* `A::B::C`
* `A::B::C::D`
* `A::B::C::D::E`

To prevent this type of cascade, you would need to assign each logger an explicit level. But the hierarchy tree is still walked to see if a logger has an explicit level or not.

Take a look at the [hierarchies example](https://github.com/TwP/logging/blob/master/examples/hierarchies.rb) and inspect your own logging hierarchy via the `Logging.show_configuration` method.

**Comment by perlun** [comment]
Hey Tim,

Thanks a lot for your thorough answers - highly appreciated. Like you suggest, creating _loads_ of loggers like I was trying is probably a very bad idea w/ this library, given the big risk for leaking memory in the long run (we are using this in a long-running http-server-based app framework).

So I think I better have to fix it to not be so dependent on "per-instance" loggers, and i.e. not change the log level in this manner re-entrantly.

It seemed pretty fundamental in terms of how the Logging library is constructed, the whole repository/locate existing logger stuff (after looking a bit at the code). I guess the log4r "disconnected logger" approach (i.e. not kept in the repository) is simply out of scope for the use case that this library was designed for?

**Comment by TwP** [comment]
Let's take a step back and ask the question "what are you trying to accomplish with a logger per instance approach?" Perhaps there is another solution.

**Comment by perlun** [comment]
> "what are you trying to accomplish with a logger per instance approach?"

The use case is this:

- I have a standard Monitor class that can be used by the individual REST resource handlers. The resource handler class is instantiated on a "one instance per page request" approach. Likewise with the Monitor, they have access to an individual Monitor instance per request.
- The Monitor has a logger. The logger name is taken from the Monitor class name.
- On a per-request basis, you can _enable_ or _disable_ logging of the metrics. This is done using a simple `_metrics` URL parameter. I.e. I can take any random web request and tack on a `&_metrics=true` parameter to the URL to enable metrics logging.

Previously, this was done by simply setting the log level _on the logger_ that is used by the Monitor. Now, with the Logging gem, that doesn't work because the Monitor logger is shared, so `level = :foo` on that logger will be very re-entrant and show errors as given here.

I already tweaked the Monitor class to keep an internal state instead, so that instead of relying on "set the log level based on input parameter", it's more now "write to the logger based on input parameter". Which is a bit bad, because I then have to add conditionals in the code, but it's not a very complex class so it's reasonably OK in this case.

If you have other suggestions for how to handle use cases like this in an even better way, I'm all ears.

**Comment by perlun** [comment]
@TwP - I felt like helping out a bit so I did a fix for the `level=` method to properly work when being called reentrantly: #132.

Of course, like has been discussed - this is not an optimal usage of the Logging gem. Still, it's always bad to raise exceptions and there may be use cases where it's not really possible to prevent the code from setting the level in multiple places, on multiple threads. So, I would consider this fix warranted.

**Comment by TwP** [comment]
Question - is this sufficiently fixed for you via #137, and can this issue be closed out? If so, please do close the issue and open up new ones with other specific thread safety issues in the library.

Thank you so much for taking the time to contribute to Logging :bow:

**Comment by perlun** [comment]
> Question - is this sufficiently fixed for you via #137, and can this issue be closed out?

I haven't had a chance to verify it, but I think so - thanks a lot for looking into this with more depth than I did (after all, you _do_ know the gem a whole lot better than me...).

Closing for now.
